### PR TITLE
✨ Upgrade CAPI to v0.4.0-beta.0

### PR DESCRIPTION
**Pull request by sbueringer**
**What this PR does / why we need it**:
Update to current CAPI version before our own beta.0 release. Also went through the CAPI v1alpha3=>v1alpha4 migration doc again (https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md).

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Partially implements: #796

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**TODOs**:

- [x] squashed commits
- if necessary:
  - [ ] includes documentation
  - [ ] adds unit tests

/hold

### CODE REVIEW

**Comment by chrischdi** [comment]
Looks like there's a different binary used in `make generate`.

/lgtm

**Comment by sbueringer** [comment]
> Looks like there's a different binary used in `make generate`.
> 
> /lgtm

Thx, fixed, ptal :)

**Comment by sbueringer** [comment]
Had to fix the cluster templates. There is a breaking change in KubeadmControlPlane (just another reason why v1alpha3 is not really compatible with v1alpha4, which is intended)

**Comment by jichenjc** [comment]
generally looks good 
let's wait for e2e test

File: hack/tools/go.mod [hunk 1]
```diff
@@ -12,10 +12,9 @@ require (
 	github.com/onsi/ginkgo v1.16.1
 	github.com/onsi/gomega v1.11.0 // indirect
 	k8s.io/apiextensions-apiserver v0.21.0 // indirect
-	k8s.io/code-generator v0.21.0
-	k8s.io/utils v0.0.0-20210305010621-2afb4311ab10 // indirect
+	k8s.io/code-generator v0.21.1
 	sigs.k8s.io/cluster-api/hack/tools v0.0.0-20210402170612-77093e2350cd
 	sigs.k8s.io/controller-tools v0.5.0
-	sigs.k8s.io/kind v0.10.0
+	sigs.k8s.io/kind v0.11.0
```
Comment (hidekazuna) [inline hack/tools/go.mod:18]: We can use v0.11.1, But It is ok when updating to v0.4.0.
https://github.com/kubernetes-sigs/kind/releases/tag/v0.11.1

File: hack/tools/go.mod [hunk 2]
```diff
@@ -12,10 +12,9 @@ require (
 	github.com/onsi/ginkgo v1.16.1
```
Comment (hidekazuna) [inline hack/tools/go.mod:12]: Should we upgrade to v1.16.4 as with go.mod? And upgrade gomega to v1.13.0 explicitly?

File: hack/tools/go.mod (hunk 1 above)
Comment (sbueringer) [reply hack/tools/go.mod]: fixed

File: hack/tools/go.mod (hunk 2 above)
Comment (sbueringer) [reply hack/tools/go.mod]: fixed

**Comment by sbueringer** [comment]
@hidekazuna thx, ptal :)
@jichenjc  last findings fixed. tests have been green once. I would like to merge relatively soon after the tests are green again after the last fix.

File: hack/tools/go.mod [hunk 3]
```diff
@@ -9,13 +9,12 @@ require (
 	github.com/google/gofuzz v1.2.0 // indirect
 	github.com/imdario/mergo v0.3.11 // indirect
 	github.com/itchyny/gojq v0.12.2
-	github.com/onsi/ginkgo v1.16.1
+	github.com/onsi/ginkgo v1.16.4
 	github.com/onsi/gomega v1.11.0 // indirect
```
Comment (hidekazuna) [inline hack/tools/go.mod:13]: gomega v1.13.0 is specified in go.mod. we can use gomega v1.13.0 here.

Comment (sbueringer) [reply hack/tools/go.mod]: I just dropped all indirect dependencies. Apparently `go mod tidy` doesn't drop them by itself.

Comment (hidekazuna) [reply hack/tools/go.mod]: Thanks updating.

**Pull request merged**